### ISSUE TITLE
Neo4J schema not inferred correctly by Neo4JGraph Object

### ISSUE THREAD

**Issue by HubertKl** [hwchase17/langchain#6380]
### System Info

langchain=0.0.2

### Who can help?

@hwchase17 

### Information

- [ ] The official example notebooks/scripts
- [ ] My own modified scripts

### Related Components

- [ ] LLMs/Chat Models
- [ ] Embedding Models
- [ ] Prompts / Prompt Templates / Prompt Selectors
- [ ] Output Parsers
- [ ] Document Loaders
- [ ] Vector Stores / Retrievers
- [ ] Memory
- [ ] Agents / Agent Executors
- [X] Tools / Toolkits
- [ ] Chains
- [ ] Callbacks/Tracing
- [ ] Async

### Reproduction

Steps to reproduce behaviors:

1. Push the following dataset to neo4J (say in neo4J browser)
```
CREATE (la:LabelA {property_a: 'a'})
CREATE (lb:LabelB {property_b1: 123, property_b2: 'b2'})
CREATE (lc:LabelC)
MERGE (la)-[:REL_TYPE]-> (lb)
MERGE (la)-[:REL_TYPE {rel_prop: 'abc'}]-> (lc)
```
2. Instantiate a Neo4JGraphObject, connect and refresh schema
```
from langchain.graphs import Neo4jGraph

graph = Neo4jGraph(
    url=NEO4J_URL,
    username=NEO4J_USERNAME,
    password=NEO4J_PASSWORD,
)

graph.refresh_schema()
print(graph.get_schema)
```
You will obtain
```
       Node properties are the following:
        [{'properties': [{'property': 'property_a', 'type': 'STRING'}], 'labels': 'LabelA'}, {'properties': [{'property': 'property_b2', 'type': 'STRING'}, {'property': 'property_b1', 'type': 'INTEGER'}], 'labels': 'LabelB'}]
        Relationship properties are the following:
        [{'type': 'REL_TYPE', 'properties': [{'property': 'rel_prop', 'type': 'STRING'}]}]
        The relationships are the following:
        ['(:LabelA)-[:REL_TYPE]->(:LabelB)']
```

### Expected behavior

```
       Node properties are the following:
        [{'properties': [{'property': 'property_a', 'type': 'STRING'}], 'labels': 'LabelA'}, {'properties': [{'property': 'property_b2', 'type': 'STRING'}, {'property': 'property_b1', 'type': 'INTEGER'}], 'labels': 'LabelB'}]
        Relationship properties are the following:
        [{'type': 'REL_TYPE', 'properties': [{'property': 'rel_prop', 'type': 'STRING'}]}]
        The relationships are the following:
        ['(:LabelA)-[:REL_TYPE]->(:LabelB)', '(:LabelA)-[:REL_TYPE]->(:LabelC)']
```

**Comment by HubertKl** [comment]
This PR fixes the query that gets the relationships in the Neo4J database. One integration test has been added too.

**Comment by jcpayne** [comment]
I'm having the same problem (Feb 2024).  My graph was created by two separate `LOAD CSV` statements.  The schema returned by Langchain Neo4jGraph apparently only sees the first import, but any Cypher query shows everything.

    node types:
        - State:
            - name: String
        - Region:
            - name: String
        - County:
            - name: String
        - Incentivename:
            - id: String
            - name: String
            - Description: String
            - Aquatic: String
            - Financialbenefit: String
            - Landuse: String
            - Level: String
            - Incentivetypebenefit: String
            - Eligibility: String
            - url: String
        - Localcontactorganization:
            - name: String
            - Localcontactnumber: String
            - Localcontactemail: String
        - Provider:
            - name: String
    
    Relationship types:
      - IS_IN [county:County,region:Region] and [region:Region,state:State]
      - SERVES [incentive:Incentivename,county:County]
      - PROVIDED_BY [incentive:Incentivename,provider:Provider]
      - SUPPORTED_LOCALLY_BY [incentive:Incentivename,localorg:Localcontactorganization]
      - OPERATES_IN [localorg:Localcontactorganization,county:County]
The first `LOAD CSV` statement loads only the County, Region, and State nodes and the 'IS_IN' property.  

Now when I try to use langchain's Neo4jGraph object:
```python
graph = Neo4jGraph(
    url="bolt://localhost:7687",
    username="neo4j",
    password="xxxx"
)
print(graph.schema)
```
Result:
Node properties are the following:
State {name: STRING},Counties {name: STRING},Region {name: STRING}
Relationship properties are the following:

The relationships are the following:
(:Counties)-[:IS_IN]->(:Region),(:Region)-[:IS_IN]->(:State)

**Comment by HubertKl** [comment]
@jcpayne 
Sorry for the silly questions but...
- Did you tried to run the python snippet before or after the second `LOAD CSV` statement? because I suspect you need to wait for the data to be fully sync to neo4j before running your python.
-  Is apoc installed on your Neo4J?
- Have you tried to run this python code instead?
```
graph = Neo4jGraph(
    url="bolt://localhost:7687",
    username="neo4j",
    password="xxxx"
)
graph.refresh_schema()
print(graph.schema)
```
- Can you please share the output for the followings:
```
print(graph.query("""
  CALL apoc.meta.data()
  YIELD label, other, elementType, type, property
  WHERE NOT type = "RELATIONSHIP" AND elementType = "node"
  WITH label AS nodeLabels, collect({property:property, type:type}) AS properties
  RETURN {labels: nodeLabels, properties: properties} AS output
""")
```
and
```
print(graph.query("""
  CALL apoc.meta.data()
  YIELD label, other, elementType, type, property
  WHERE NOT type = "RELATIONSHIP" AND elementType = "relationship"
  WITH label AS nodeLabels, collect({property:property, type:type}) AS properties
  RETURN {type: nodeLabels, properties: properties} AS output
""")
```
and 

```
print(graph.query("""
  CALL apoc.meta.data()
  YIELD label, other, elementType, type, property
  WHERE type = "RELATIONSHIP" AND elementType = "node"
  UNWIND other AS other_node
  RETURN {start: label, type: property, end: toString(other_node)} AS output
""")
```
- Can you also please run the same Cypher queries directly in Neo4j and not in python for comparison please?
- Can you please provide a way to reproduce the bug with a minimal sample dummy data to sync to neo4j please?

**Comment by HubertKl** [comment]
@jcpayne 
Also ...
- Can you please give the versions of langchain, Neo4J and APOC that you are using please

**Comment by jcpayne** [comment]
Thank you for the quick response.  First, I am a newcomer to Neo4j so it's certainly possible that I have messed something up.  

**Versions**
I'm using Neo4j Desktop version 5.12 (installed about a week ago) with APOC version 5.12.  `pip show langchain` says that I'm using langchain 0.1.4.  

**Your questions**: 
- Did you tried to run the python snippet before or after the second LOAD CSV statement? because I suspect you need to wait for the data to be fully sync to neo4j before running your python. ANSWER: Well after, and it is a small database so it had fully synced.  Also, the Neo4j browser shows the correct results with all of the nodes and relationships.

- Is apoc installed on your Neo4J? YES

- Have you tried to run this python code ... instead?  -->SAME RESULT

**The queries you asked me to run**
To save space, I'll refer to your queries as Q1, Q2 and Q3 (in order).
Q1 run in Python.  Result: it misses nodes from the second `LOAD CSV` statement.
```
[{'output': {'labels': 'State', 'properties': [{'property': 'name', 'type': 'STRING'}]}}, {'output': {'labels': 'Counties', 'properties': [{'property': 'name', 'type': 'STRING'}]}}, {'output': {'labels': 'Region', 'properties': [{'property': 'name', 'type': 'STRING'}]}}]
```
Q1 Run directly in Neo4j Desktop browser.  Result: now it finds all of the nodes.
```
{
  "labels": "State",
  "properties": [
    {
      "property": "name",
      "type": "STRING"
    }
  ]
}
{
  "labels": "Region",
  "properties": [
    {
      "property": "name",
      "type": "STRING"
    }
  ]
}
{
  "labels": "Counties",
  "properties": [
    {
      "property": "name",
      "type": "STRING"
    }
  ]
}
{
  "labels": "Incentivename",
  "properties": [
    {
      "property": "Landuse",
      "type": "STRING"
    },
    {
      "property": "Aquatic",
      "type": "STRING"
    },
    {
      "property": "Incentivetypebenefit",
      "type": "STRING"
    },
    {
      "property": "Financialbenefit",
      "type": "STRING"
    },
    {
      "property": "Level",
      "type": "STRING"
    },
    {
      "property": "url",
      "type": "STRING"
    },
    {
      "property": "name",
      "type": "STRING"
    },
    {
      "property": "id",
      "type": "STRING"
    },
    {
      "property": "Description",
      "type": "STRING"
    },
    {
      "property": "Eligibility",
      "type": "STRING"
    }
  ]
}
{
  "labels": "Provider",
  "properties": [
    {
      "property": "name",
      "type": "STRING"
    }
  ]
}
{
  "labels": "Localcontactorganization",
  "properties": [
    {
      "property": "Localcontactnumber",
      "type": "STRING"
    },
    {
      "property": "name",
      "type": "STRING"
    },
    {
      "property": "Localcontactemail",
      "type": "STRING"
    }
  ]
}
```
Q2 run in Python.  Result: returns an empty set
```
[] 
```
Q2 run directly in Neo4j.  Result: now it finds all of the correct relationships:
```
{
  "start": "Region",
  "type": "IS_IN",
  "end": "State"
}
{
  "start": "Counties",
  "type": "IS_IN",
  "end": "Region"
}
{
  "start": "Incentivename",
  "type": "SERVES",
  "end": "Counties"
}
{
  "start": "Incentivename",
  "type": "PROVIDED_BY",
  "end": "Provider"
}
{
  "start": "Incentivename",
  "type": "SUPPORTED_LOCALLY_BY",
  "end": "Localcontactorganization"
}
{
  "start": "Localcontactorganization",
  "type": "OPERATES_IN",
  "end": "Counties"
}
```
Q3 run in Python.  Result: misses all of the relationships added in the second `LOAD CSV` statement.
```
[{'output': {'start': 'Counties', 'type': 'IS_IN', 'end': 'Region'}}, {'output': {'start': 'Region', 'type': 'IS_IN', 'end': 'State'}}]
```
Q3 run directly in Neo4j Desktop browser.  Result: the list of relationships is complete.
```
{
  "start": "Region",
  "type": "IS_IN",
  "end": "State"
}
{
  "start": "Counties",
  "type": "IS_IN",
  "end": "Region"
}
{
  "start": "Incentivename",
  "type": "SERVES",
  "end": "Counties"
}
{
  "start": "Incentivename",
  "type": "PROVIDED_BY",
  "end": "Provider"
}
{
  "start": "Incentivename",
  "type": "SUPPORTED_LOCALLY_BY",
  "end": "Localcontactorganization"
}
{
  "start": "Localcontactorganization",
  "type": "OPERATES_IN",
  "end": "Counties"
}
```
As you suggested, I will work on creating a sample database but it will take some experimentation to see if I can get something that creates the same errors, so in the meantime I will post this comment as-is.  Please let me know if there is anything else I can provide that would be useful in the meantime.  I appreciate you taking the time to look at it.

**Comment by HubertKl** [comment]
@jcpayne 
I can't reproduce the bug, with the same version of langchain, neo4j and APOC you mentioned. I tried the following,: created 2 CSV file, load them into neo4j in 2 "LOAD CSV" queries, run python and I have what I expected from it.

In more details I created a `artists.csv` file:
```
1,ABBA,1992
2,Roxette,1986
3,Europe,1979
4,The Cardigans,1992
```
and a `color.csv` file
```
1,ABBA,blue
2,Roxette,blue
3,Europe,red
4,The Cardigans,orange
```
then run the queries
```
LOAD CSV FROM 'file:///artists.csv' AS line
MERGE (artist:Artist {name: line[1]})
MERGE (year:Year {year: toInteger(line[2])})
MERGE (artist)-[:IN_YEAR]->(year)
```
and
```
LOAD CSV FROM 'file:///color.csv' AS line
MERGE (artist:Artist {name: line[1]})
MERGE (color:Color {color: line[2]})
MERGE (artist)-[:HAS_COLOR]->(color)
```
after the load was done, I ran in python:
```
from langchain.graphs.neo4j_graph import Neo4jGraph

graph = Neo4jGraph(
    url="bolt://localhost:7687",
    username=USERNAME,
    password=PASSWORD,
)

graph.refresh_schema()
print(graph.get_schema)
```
And I got what I expected from it, i.e.
```
Node properties are the following:
Artist {name: STRING},Year {year: INTEGER},Color {color: STRING}
Relationship properties are the following:

The relationships are the following:
(:Artist)-[:HAS_COLOR]->(:Color),(:Artist)-[:IN_YEAR]->(:Year)
```

So unless I am missing something, I believe this is your Neo4J setup that might be the source of your problem. I would suspect that accidentally your are running 2 instances of neo4j or 2 databases, one with only one LOAD CSV statement completed the other with both LOAD CSV completed.

In Neo4J desktop, please go to the DBMS tab and see if you have several databases active. By default the `Neo4jGraph` object will connect will connect to the `neo4j` database. You can change that by init the `Neo4jGraph` object like this 
```
graph = Neo4jGraph(
    url="bolt://localhost:7687",
    username=USERNAME,
    password=PASSWORD,
    database="DATABASE_NAME_HERE, by default neo4j",
)
```

If you only have "neo4j" and "system" as available database, then you are likely running 2 instances of Neo4J. Close all neo4j instances, restart your computer and after reboot start only one instance of neo4j.

I might be missing something though... Please let me know if this help or if you can find a way to reproduce your bug.

**Comment by jcpayne** [comment]
Oh I feel foolish!  That solved it.  Thank you so much for helping to pinpoint the problem.  In fact, I think it may have been an even simpler problem: langchain was just connecting to the default `Neo4j` database, rather than a new database I had created.  I forgot that I had added a few nodes and relationships to the default `Neo4j` database when I was trying it out for the first time.

Several posts online say that it's only possible to run one instance of neo4j at a time, and I'm not sure if that happened or not.  

- Before rebooting, I checked and:
- only one DBMS was active.  It was one I created, called CCLC-DBMS.  
- In it were three databases: 1) `Neo4j (default)`, 2) the new database I was working on, called `cclc-db`, and 3) `system` (as far as I can tell, the default and system databases can't be deleted  because the option is grayed out).  
- The Neo4j browser prompt showed `cclc-db`.  

In short, I think the Neo4j side of things was all running correctly: it looked to me like only one instance of Neo4j was running, only one Neo4j browser was running, and the correct DBMS and database were running.  

Adding `database=cclc-db` to the init of the `Neo4j` graph object in my Python script and then rebooting fixed the problem.  

Two questions: 1) Is there any way to tell in future if I have more than one instance of Neo4j going? and 2) I started with a slightly different approach using the `Neo4jGPTQuery` object, and its signature doesn't allow for changing the database name.  Is there a way to feed it a different database (or is it better practice to use the default db in general)?

**Comment by HubertKl** [comment]
Great!

To check if you have several instance of neo4j running, look at system monitor, or if you use docker, use `docker ps`. 

I don't know the `Neo4jGPTQuery` object, look at the documentation or the source code, here may be more details on it.

Well with neo4j this is a bit special.  In theory you wish to connect all your wonderful data nodes of your wonderful data with wonderful relationships in one wonderful fully connected graph... In practice this is fairly convenient to have a bit of of order and separation of databases... It used to be that Neo4J community edition allowed only for only the default databases (`neo4j` and `system`) and not to create any new ones.
